src/codegen/sys/lib_.rs

```diff
@@ -71,17 +71,19 @@ fn generate_lib(w: &mut dyn Write, env: &Env) -> Result<()> {
     generate_classes_structs(w, env, &classes)?;
     generate_interfaces_structs(w, env, &interfaces)?;
 
-    write_link_attr(w, &env.namespaces.main().shared_libs)?;
-    writeln!(w, "extern \"C\" {{")?;
-    functions::generate_enums_funcs(w, env, &enums)?;
-    functions::generate_bitfields_funcs(w, env, &bitfields)?;
-    functions::generate_unions_funcs(w, env, &unions)?;
-    functions::generate_records_funcs(w, env, &records)?;
-    functions::generate_classes_funcs(w, env, &classes)?;
-    functions::generate_interfaces_funcs(w, env, &interfaces)?;
-    functions::generate_other_funcs(w, env, &ns.functions)?;
-
-    writeln!(w, "\n}}")?;
+    if !env.namespaces.main().shared_libs.is_empty() {
```

Comment:
You probably want to add something here that makes things fail if there are any functions but no library name

Reply:
Something like this?

```rust
    if !ns.functions.is_empty() && env.namespaces.main().shared_libs.is_empty() {
        return Err(Error::new(ErrorKind::Other, "No shared library found, but functions found"));
    }
```

Reply:
Not sure what the easiest. `ns.functions` only contains the global functions and not the methods IIRC.

Reply:
Maybe something like this:
```rust
    if !ns.package_names.is_empty() && env.namespaces.main().shared_libs.is_empty() {
        return Err(Error::new(ErrorKind::Other, "No shared library found, but package names found"));
    }
```
That way `.gir`s with libraries defined have to have a package name, but headers like this one need to not have one. That way you can't have the combination of having a shared library with functions, but not defining the package.

Reply:
Having a package name seems to be orthogonal from have shared libraries. Your header-only library will have a package name (and as such a `pkg-config` file) or not?

I think what just needs checking here is whether there are global functions, or enums/object/flags/records have functions

Reply:
Pushed with it modified to this, which checks all the function lengths. Adding more than just `classes` and `interfaces` might be a bit overkill though.

```rust
    if env.namespaces.main().shared_libs.is_empty()
        && !(records.iter().all(|x| x.functions.is_empty())
            && classes.iter().all(|x| x.functions.is_empty())
            && interfaces.iter().all(|x| x.functions.is_empty())
            && bitfields.iter().all(|x| x.functions.is_empty())
            && enums.iter().all(|x| x.functions.is_empty())
            && unions.iter().all(|x| x.functions.is_empty()))
    {
        return Err(Error::new(
            ErrorKind::Other,
            "No shared library found, but functions were found",
        ));
    }
```